# Exploring Gun Deaths in the US

The dataset contains data on gun deaths occuring in the USA from 2012 to 2014. Each row represents one fatality, with the following column structure:

- `row identifier` - record count
- `year` - year of incident
- `month` - month of incident
- `intent` - the intent of the perpetrator of the crime. Possible fields include: Suicide, Accidental, NA, Homicide, Undetermined
- `police` - whether police was involved: 1 (true), 0 (false)
- `sex` - the gender of the victim
- `age` - victim's age
- `race` - victim's race: 
>- Asian/Pacific Islander 
>- Native American/Native Alaskan 
>- Black 
>- Hispanic 
>- White

- `hispanic` - a code indicating Hispanic origin of the Victim
- `place` - where the shooting occured
- `education` - victim's level of education:
>- 1 - Less than High School
>- 2 - Graduated from High School or equivalent
>- 3 - Some College
>- 4 - At least graduated from College
>- 5 - Not available 


In [1]:
# read in csv file and parse header from data
import csv
f = open("gun-data.csv","r")
data = list(csv.reader(f))
print(data[:5])

[['', 'year', 'month', 'intent', 'police', 'sex', 'age', 'race', 'hispanic', 'place', 'education'], ['1', '2012', '01', 'Suicide', '0', 'M', '34', 'Asian/Pacific Islander', '100', 'Home', 'BA+'], ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', 'Some college'], ['3', '2012', '01', 'Suicide', '0', 'M', '60', 'White', '100', 'Other specified', 'BA+'], ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', 'BA+']]


In [2]:
headers = data[0]
data = data[1:]
print(headers)
print(data[:5])

['', 'year', 'month', 'intent', 'police', 'sex', 'age', 'race', 'hispanic', 'place', 'education']
[['1', '2012', '01', 'Suicide', '0', 'M', '34', 'Asian/Pacific Islander', '100', 'Home', 'BA+'], ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', 'Some college'], ['3', '2012', '01', 'Suicide', '0', 'M', '60', 'White', '100', 'Other specified', 'BA+'], ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', 'BA+'], ['5', '2012', '02', 'Suicide', '0', 'M', '31', 'White', '100', 'Other specified', 'HS/GED']]


In [3]:
# count number of deaths per year
years = [row[1] for row in data]
year_counts = {}
for y in years:
    if y not in year_counts:
        year_counts[y] = 1
    else:
        year_counts[y] += 1
year_counts

{'2012': 33563, '2013': 33636, '2014': 33599}

In [4]:
# convert year/month columns to datetime objects
import datetime as d
dates = [d.datetime(year=int(row[1]), month=int(row[2]), day=1) for row in data]
print(dates[:5])

[datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 2, 1, 0, 0), datetime.datetime(2012, 2, 1, 0, 0)]


In [5]:
date_counts = {}
for date in dates:
    if date not in date_counts:
        date_counts[date] = 1
    else:
        date_counts[date] += 1
date_counts


{datetime.datetime(2012, 1, 1, 0, 0): 2758,
 datetime.datetime(2012, 2, 1, 0, 0): 2357,
 datetime.datetime(2012, 3, 1, 0, 0): 2743,
 datetime.datetime(2012, 4, 1, 0, 0): 2795,
 datetime.datetime(2012, 5, 1, 0, 0): 2999,
 datetime.datetime(2012, 6, 1, 0, 0): 2826,
 datetime.datetime(2012, 7, 1, 0, 0): 3026,
 datetime.datetime(2012, 8, 1, 0, 0): 2954,
 datetime.datetime(2012, 9, 1, 0, 0): 2852,
 datetime.datetime(2012, 10, 1, 0, 0): 2733,
 datetime.datetime(2012, 11, 1, 0, 0): 2729,
 datetime.datetime(2012, 12, 1, 0, 0): 2791,
 datetime.datetime(2013, 1, 1, 0, 0): 2864,
 datetime.datetime(2013, 2, 1, 0, 0): 2375,
 datetime.datetime(2013, 3, 1, 0, 0): 2862,
 datetime.datetime(2013, 4, 1, 0, 0): 2798,
 datetime.datetime(2013, 5, 1, 0, 0): 2806,
 datetime.datetime(2013, 6, 1, 0, 0): 2920,
 datetime.datetime(2013, 7, 1, 0, 0): 3079,
 datetime.datetime(2013, 8, 1, 0, 0): 2859,
 datetime.datetime(2013, 9, 1, 0, 0): 2742,
 datetime.datetime(2013, 10, 1, 0, 0): 2808,
 datetime.datetime(2013, 11,

In [6]:
sex_counts = {}
for row in data:
    if row[5] not in sex_counts:
        sex_counts[row[5]] = 1
    else:
        sex_counts[row[5]] += 1
sex_counts


{'F': 14449, 'M': 86349}

In [7]:
race_counts = {}
for row in data:
    if row[7] not in race_counts:
        race_counts[row[7]] = 1
    else:
        race_counts[row[7]] += 1
race_counts

{'Asian/Pacific Islander': 1326,
 'Black': 23296,
 'Hispanic': 9022,
 'Native American/Native Alaskan': 917,
 'White': 66237}

# Analysis Pt. 1: Race vs. Sex
- Incident counts remain relatively consistent by day, generally averaging ~2,700
- Victims of gun violence are predominantly male
- Whites and Blacks comprise the majority of victims based on total counts

In [8]:
f = open("census.csv",'r')
census = list(csv.reader(f))
census = census[:2]
census

[['',
  'Id',
  'Year',
  'Id.1',
  'Sex',
  'Id.2',
  'Hispanic Origin',
  'Id.3',
  'Id2',
  'Geography',
  'Total',
  'Race Alone - White',
  'Race Alone - Hispanic',
  'Race Alone - Black or African American',
  'Race Alone - American Indian and Alaska Native',
  'Race Alone - Asian',
  'Race Alone - Native Hawaiian and Other Pacific Islander',
  'Two or More Races'],
 ['0',
  'cen42010',
  'April 1, 2010 Census',
  'totsex',
  'Both Sexes',
  'tothisp',
  'Total',
  '0100000US',
  'NaN',
  'United States',
  '308745538',
  '197318956',
  '44618105',
  '40250635',
  '3739506',
  '15159516',
  '674625',
  '6984195']]

In [9]:
mapping = {}
for i in range(-7,0):
    census[1][i] = int(census[1][i])

mapping["White"] = census[1][-7]
mapping["Hispanic"] = census[1][-6]
mapping["Black"] = census[1][-5]
mapping["Native American/Native Alaskan"] = census[1][-4]
mapping["Asian/Pacific Islander"] = census[1][-2] + census[1][-3]
mapping


{'Asian/Pacific Islander': 15834141,
 'Black': 40250635,
 'Hispanic': 44618105,
 'Native American/Native Alaskan': 3739506,
 'White': 197318956}

In [10]:
race_per_hundredk = {}
for k, v in race_counts.items():
    race_per_hundredk[k] = round(v / mapping[k] * 100000, 2)
race_per_hundredk

{'Asian/Pacific Islander': 8.37,
 'Black': 57.88,
 'Hispanic': 20.22,
 'Native American/Native Alaskan': 24.52,
 'White': 33.57}

In [11]:
intents = [row[3] for row in data]
races = [row[7] for row in data]
print(intents[:5])
print(races[:5])

homicide_race_counts = {}
for i, race in enumerate(races):
    if intents[i] == "Homicide":
        if race not in homicide_race_counts:
            homicide_race_counts[race] = 1
        else:
            homicide_race_counts[race] += 1
homicide_race_counts

['Suicide', 'Suicide', 'Suicide', 'Suicide', 'Suicide']
['Asian/Pacific Islander', 'White', 'White', 'White', 'White']


{'Asian/Pacific Islander': 559,
 'Black': 19510,
 'Hispanic': 5634,
 'Native American/Native Alaskan': 326,
 'White': 9147}

In [12]:
for k, v in homicide_race_counts.items():
    homicide_race_counts[k] = round(v / mapping[k] * 100000, 2)
homicide_race_counts

{'Asian/Pacific Islander': 3.53,
 'Black': 48.47,
 'Hispanic': 12.63,
 'Native American/Native Alaskan': 8.72,
 'White': 4.64}

# Analysis Pt. 2: Gun Death Rates by Race
- Based on total counts, it looks like Whites are most susceptible to gun violence.
- After examining each race on the same scale (per 100,000), the data shows that Blacks (48.5 per 100k) are most likely to be victims of gun violence compared to other races.
- Looking at Homicides only, this disparity between Blacks and other races becomes more apparent. Homicide incidents are 12x more likely to occur to Black people than White people (4.6 per 100k) and 4x more likely than Hispanics (12.6 per 100k), the race with the second highest homicide rates.

## Next Steps:
1. Identify how other intents of gun violence incidents vary by race:
> - Ratios of each intent per race
> - Rates per hundred thousand of each intent

2. Analyze trends in gun deaths over time
> - Total gun deaths per year/month
> - Total gun deaths per year/month categorized by intent
> - Total gun deaths per year/month categorized by race

3. Analyze correlations in education for groups with the highest gun deaths

In [13]:
def rates_per_100k_race(intent):
    intents = [row[3] for row in data]
    races = [row[7] for row in data]
    race_counts = {}
    for i, race in enumerate(races):
        if intents[i] == intent:
            if race not in race_counts:
                race_counts[race] = 1
            else:
                race_counts[race] += 1
    for k, v in race_counts.items():
        race_counts[k] = round(v / mapping[k] * 100000, 2)
    return race_counts

accidental_race_counts = rates_per_100k_race("Accidental")
suicide_race_counts = rates_per_100k_race("Suicide")

In [14]:
accidental_race_counts

{'Asian/Pacific Islander': 0.08,
 'Black': 0.81,
 'Hispanic': 0.32,
 'Native American/Native Alaskan': 0.59,
 'White': 0.57}

In [15]:
suicide_race_counts

{'Asian/Pacific Islander': 4.71,
 'Black': 8.28,
 'Hispanic': 7.11,
 'Native American/Native Alaskan': 14.84,
 'White': 28.06}

In [16]:
def intent_ratios(race):
    intents = [row[3] for row in data]
    races = [row[7] for row in data]
    
    # count gun incidents for each intent under given race input
    counts = {}

    for i, intent in enumerate(intents):
        if races[i] == race:
            if intent not in counts:
                counts[intent] = 1
            else:
                counts[intent] += 1
    
    # calculate total counts per race in gun data
    race_counts = {}
    for row in data:
        if row[7] not in race_counts:
            race_counts[row[7]] = 1
        else:
            race_counts[row[7]] += 1
    
    # calculate ratios of each intent
    for k, v in counts.items():
        counts[k] = round(v / race_counts[race] * 100, 2)
    
    counts[".Race"] = race
    return counts


intent_ratios = [intent_ratios(race) for race in set(races)]
intent_ratios


[{'.Race': 'Black',
  'Accidental': 1.41,
  'Homicide': 83.75,
  'Suicide': 14.3,
  'Undetermined': 0.54},
 {'.Race': 'White',
  'Accidental': 1.71,
  'Homicide': 13.81,
  'NA': 0.0,
  'Suicide': 83.6,
  'Undetermined': 0.88},
 {'.Race': 'Native American/Native Alaskan',
  'Accidental': 2.4,
  'Homicide': 35.55,
  'Suicide': 60.52,
  'Undetermined': 1.53},
 {'.Race': 'Hispanic',
  'Accidental': 1.61,
  'Homicide': 62.45,
  'Suicide': 35.15,
  'Undetermined': 0.8},
 {'.Race': 'Asian/Pacific Islander',
  'Accidental': 0.9,
  'Homicide': 42.16,
  'Suicide': 56.18,
  'Undetermined': 0.75}]

In [17]:
# convert year columns to datetime objects
import datetime as d
dates = [d.datetime(year=int(row[1]), month=1, day=1) for row in data]
print(dates[:5])

# calculate total gun death counts by year
date_counts = {}
for date in dates:
    if date not in date_counts:
        date_counts[date] = 1
    else:
        date_counts[date] += 1
date_counts

[datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 1, 1, 0, 0)]


{datetime.datetime(2012, 1, 1, 0, 0): 33563,
 datetime.datetime(2013, 1, 1, 0, 0): 33636,
 datetime.datetime(2014, 1, 1, 0, 0): 33599}

In [18]:
# calculates gun death counts by intent per unit: month or year
# defaults to month if no unit given
def intent_date_counts(intent, unit="year/month"):
    import datetime as d
    dates = []
    if unit == "year":
        dates = [d.datetime(year=int(row[1]), month=1, day=1) for row in data]
    else:
        dates = [d.datetime(year=1, month=int(row[2]), day=1) for row in data]        
    
    intents = [row[3] for row in data]
    
    # count number of gun deaths by intent based on enumerated lists
    date_counts = {}
    for i, date in enumerate(dates):
        if intents[i] == intent:
            if date not in date_counts:
                date_counts[date] = 1
            else:
                date_counts[date] += 1
    
    # create sorted lsit for representation
    date_counts = sorted([(k, v) for k, v in date_counts.items()])
    date_counts.insert(0,"Intent: " + intent)
    
    return date_counts

intent_per_year = [intent_date_counts(intent, "year") for intent in set(intents)]
intent_per_year

[['Intent: Undetermined',
  (datetime.datetime(2012, 1, 1, 0, 0), 256),
  (datetime.datetime(2013, 1, 1, 0, 0), 281),
  (datetime.datetime(2014, 1, 1, 0, 0), 270)],
 ['Intent: Accidental',
  (datetime.datetime(2012, 1, 1, 0, 0), 548),
  (datetime.datetime(2013, 1, 1, 0, 0), 505),
  (datetime.datetime(2014, 1, 1, 0, 0), 586)],
 ['Intent: NA', (datetime.datetime(2013, 1, 1, 0, 0), 1)],
 ['Intent: Homicide',
  (datetime.datetime(2012, 1, 1, 0, 0), 12093),
  (datetime.datetime(2013, 1, 1, 0, 0), 11674),
  (datetime.datetime(2014, 1, 1, 0, 0), 11409)],
 ['Intent: Suicide',
  (datetime.datetime(2012, 1, 1, 0, 0), 20666),
  (datetime.datetime(2013, 1, 1, 0, 0), 21175),
  (datetime.datetime(2014, 1, 1, 0, 0), 21334)]]

In [19]:
intent_per_month = [intent_date_counts(intent, "month") for intent in set(intents)]
intent_per_month

[['Intent: Undetermined',
  (datetime.datetime(1, 1, 1, 0, 0), 72),
  (datetime.datetime(1, 2, 1, 0, 0), 56),
  (datetime.datetime(1, 3, 1, 0, 0), 66),
  (datetime.datetime(1, 4, 1, 0, 0), 73),
  (datetime.datetime(1, 5, 1, 0, 0), 72),
  (datetime.datetime(1, 6, 1, 0, 0), 68),
  (datetime.datetime(1, 7, 1, 0, 0), 57),
  (datetime.datetime(1, 8, 1, 0, 0), 73),
  (datetime.datetime(1, 9, 1, 0, 0), 81),
  (datetime.datetime(1, 10, 1, 0, 0), 51),
  (datetime.datetime(1, 11, 1, 0, 0), 78),
  (datetime.datetime(1, 12, 1, 0, 0), 60)],
 ['Intent: Accidental',
  (datetime.datetime(1, 1, 1, 0, 0), 152),
  (datetime.datetime(1, 2, 1, 0, 0), 127),
  (datetime.datetime(1, 3, 1, 0, 0), 134),
  (datetime.datetime(1, 4, 1, 0, 0), 98),
  (datetime.datetime(1, 5, 1, 0, 0), 115),
  (datetime.datetime(1, 6, 1, 0, 0), 112),
  (datetime.datetime(1, 7, 1, 0, 0), 149),
  (datetime.datetime(1, 8, 1, 0, 0), 164),
  (datetime.datetime(1, 9, 1, 0, 0), 118),
  (datetime.datetime(1, 10, 1, 0, 0), 131),
  (datetime.

In [20]:
# calculates gun death counts by race per unit: month or year
# defaults to month if no unit given
def race_date_counts(race, unit="month"):
    import datetime as d
    dates = []
    if unit == "year":
        dates = [d.datetime(year=int(row[1]), month=1, day=1) for row in data]
    else:
        dates = [d.datetime(year=1, month=int(row[2]), day=1) for row in data]        
    
    races = [row[7] for row in data]

    # count number of gun deaths by intent based on enumerated lists
    date_counts = {}
    for i, date in enumerate(dates):
        if races[i] == race:
            if date not in date_counts:
                date_counts[date] = 1
            else:
                date_counts[date] += 1
    
    # create sorted lsit for representation
    date_counts = sorted([(k, v) for k, v in date_counts.items()])
    date_counts.insert(0,"Race: " + race)
    
    return date_counts

counts_per_year_by_race = [race_date_counts(race, "year") for race in set(races)]
counts_per_year_by_race

[['Race: Black',
  (datetime.datetime(2012, 1, 1, 0, 0), 7946),
  (datetime.datetime(2013, 1, 1, 0, 0), 7712),
  (datetime.datetime(2014, 1, 1, 0, 0), 7638)],
 ['Race: White',
  (datetime.datetime(2012, 1, 1, 0, 0), 21793),
  (datetime.datetime(2013, 1, 1, 0, 0), 22222),
  (datetime.datetime(2014, 1, 1, 0, 0), 22222)],
 ['Race: Native American/Native Alaskan',
  (datetime.datetime(2012, 1, 1, 0, 0), 310),
  (datetime.datetime(2013, 1, 1, 0, 0), 296),
  (datetime.datetime(2014, 1, 1, 0, 0), 311)],
 ['Race: Hispanic',
  (datetime.datetime(2012, 1, 1, 0, 0), 3061),
  (datetime.datetime(2013, 1, 1, 0, 0), 2951),
  (datetime.datetime(2014, 1, 1, 0, 0), 3010)],
 ['Race: Asian/Pacific Islander',
  (datetime.datetime(2012, 1, 1, 0, 0), 453),
  (datetime.datetime(2013, 1, 1, 0, 0), 455),
  (datetime.datetime(2014, 1, 1, 0, 0), 418)]]

In [21]:
counts_per_month_by_race = [race_date_counts(race, "month") for race in set(races)]
counts_per_month_by_race

[['Race: Black',
  (datetime.datetime(1, 1, 1, 0, 0), 1887),
  (datetime.datetime(1, 2, 1, 0, 0), 1415),
  (datetime.datetime(1, 3, 1, 0, 0), 1797),
  (datetime.datetime(1, 4, 1, 0, 0), 1921),
  (datetime.datetime(1, 5, 1, 0, 0), 2035),
  (datetime.datetime(1, 6, 1, 0, 0), 2057),
  (datetime.datetime(1, 7, 1, 0, 0), 2220),
  (datetime.datetime(1, 8, 1, 0, 0), 2071),
  (datetime.datetime(1, 9, 1, 0, 0), 1966),
  (datetime.datetime(1, 10, 1, 0, 0), 1954),
  (datetime.datetime(1, 11, 1, 0, 0), 1884),
  (datetime.datetime(1, 12, 1, 0, 0), 2089)],
 ['Race: White',
  (datetime.datetime(1, 1, 1, 0, 0), 5462),
  (datetime.datetime(1, 2, 1, 0, 0), 4910),
  (datetime.datetime(1, 3, 1, 0, 0), 5607),
  (datetime.datetime(1, 4, 1, 0, 0), 5608),
  (datetime.datetime(1, 5, 1, 0, 0), 5741),
  (datetime.datetime(1, 6, 1, 0, 0), 5634),
  (datetime.datetime(1, 7, 1, 0, 0), 5770),
  (datetime.datetime(1, 8, 1, 0, 0), 5668),
  (datetime.datetime(1, 9, 1, 0, 0), 5621),
  (datetime.datetime(1, 10, 1, 0, 0), 

# Analysis Pt. 3: Conclusion

## Gun deaths categorized by intent
- After analyzing gun deaths by intent, the data shows that **Blacks have the highest rate of Accidental gun deaths (.81) and Homicide gun deaths (48.47) per 100,000; while Whites have the highest rate of Suicide gun deaths (28.3) per 100,000.**
- These findings are further supported after looking at intent ratios by population as shown below. **83.75% of gun deaths among Blacks are Homicide related whereas 83.6% of gun deaths among Whites are Suicide related.**
- Intent ratios by population show us that most gun deaths for each race (~95%) are either Homicide or Suicide related. Asians and Native American/Native Alaskans are more likely to be involved in Suicide related cases whereas Hispanics are more likely to be involved in Homicide related cases.

>  **Race: Black**
  'Accidental': 1.41,
  'Homicide': 83.75,
  'Suicide': 14.3,
  'Undetermined': 0.54
  
>  **Race: White**
  'Accidental': 1.71,
  'Homicide': 13.81,
  'NA': 0.0,
  'Suicide': 83.6,
  'Undetermined': 0.88
  
>  **Race: Asian/Pacific Islander**
  'Accidental': 0.9,
  'Homicide': 42.16,
  'Suicide': 56.18,
  'Undetermined': 0.75}
  
> **Race: Native American/Native Alaskan**
  'Accidental': 2.4,
  'Homicide': 35.55,
  'Suicide': 60.52,
  'Undetermined': 1.53
  
> **Race: Hispanic**
  'Accidental': 1.61,
  'Homicide': 62.45,
  'Suicide': 35.15,
  'Undetermined': 0.8
  
## Gun deaths categorized by dates
- After analyzing gun deaths by year and intent, two things become evident:
    1. **Homicide counts have slowly declined from 2012 to 2014.** While we don't have corresponding census data to show that rates per 100,000 are also decreasing, it might be safe to assume that rates are also decreasing since population has generally been increasing.
    2. **Suicide counts have slowly increased from 2012 to 2014.** However, due to a lack of census data during these years, it's uncertain as to whether or not counts have increased along with popuation growth, or have actually increased due to higher suicide rates.

> **Homicide gun deaths per year**
   -  2012: 12,093
   -  2013: 11,674
   -  2014: 11,409
  
> **Suicide gun deaths per year**
  -  2012: 20,666
  -  2013: 21,175
  -  2014: 21,334
